### PR TITLE
galaxy-galaxy lensing simulations

### PR DESCRIPTION
**Pull request by sibirrer**
This is an initial skeleton pipeline with modular classes dedicated to lens, source, and the combined lensing system.
An example class to render (realistic) galaxy populations are coming from the skypy library with SED's that are then passed through the LSST bands.

The modularity should allow to replace the galaxy library, change the deflector models, line-of-sight descriptions etc.

Separately, there are routines to simulate realizations of the lenses. At the moment there is an example translating the lens description into arguments understandable by lenstronomy and then images being simulated. Other features will then translate the lensing arguments to quantities understood by butlerGen3 etc.

Tests are currently mostly lacking and I will update the pull request.
Part of the contributions are made by @trivialTZ.

A notebook is added to demonstrate the first skeleton version of the pipeline.
Any feedback is welcomed!

### CODE REVIEW

**Comment by drphilmarshall** [comment]
Great stuff! Really nice to see this at the LSST SL pipeline meeting this morning, thanks @sibirrer :-) 

It looks like the idea is to illustrate the `sim-pipeline` package functionality with a set of notebooks (of which this is the first), so that a new user/contributor can quickly learn how the package works and start coding. I think this is a very good way to go! My main suggestion is that we include some markdown cells in each notebook to outline the scope of the tutorial in the form of a set of learning objectives, explain the setup needed to run the tutorial, and then explain what is going on in each section (with cross-links to the code and other documentation as needed). The [Rubin Stack Club tutorial template](https://github.com/LSSTScienceCollaborations/StackClub/blob/master/GettingStarted/templates/template_Notebook.ipynb) might be a good design to follow, perhaps?

Adding/editing markdown cells is something I think new/aspiring contributors like me could start helping with straight away, as we get started with the package. For the gg lensing tutorial, would you want those markdown cells via this PR, or a separate one?

**Comment by sibirrer** [comment]
> Great stuff! Really nice to see this at the LSST SL pipeline meeting this morning, thanks @sibirrer :-)
> 
> It looks like the idea is to illustrate the `sim-pipeline` package functionality with a set of notebooks (of which this is the first), so that a new user/contributor can quickly learn how the package works and start coding. I think this is a very good way to go! My main suggestion is that we include some markdown cells in each notebook to outline the scope of the tutorial in the form of a set of learning objectives, explain the setup needed to run the tutorial, and then explain what is going on in each section (with cross-links to the code and other documentation as needed). The [Rubin Stack Club tutorial template](https://github.com/LSSTScienceCollaborations/StackClub/blob/master/GettingStarted/templates/template_Notebook.ipynb) might be a good design to follow, perhaps?
> 
> Adding/editing markdown cells is something I think new/aspiring contributors like me could start helping with straight away, as we get started with the package. For the gg lensing tutorial, would you want those markdown cells via this PR, or a separate one?

Thanks @drphilmarshall ! Indeed, the documentation in the jupyter notebook is missing. If you can do it directly in the PR, that would be great, or leave comments on the NB review. I can also simply update the notebook with a first batch of MD text before a further review about the documentation

**Comment by sibirrer** [comment]
@nkhadka21 can you instead make pull request from your fork to the gg_lensing branch instead of direct pushes to the branch. Thank you!

File: sim_pipeline/RSP.py
Comment (sibirrer) [inline sim_pipeline/RSP.py:1]: file names should be lower case letters

File: sim_pipeline/RSP.py
Comment (sibirrer) [inline sim_pipeline/RSP.py:13]: is galsim a dependence of the sim-pipeline package? If not, please add it, as well as all the other packages that are imported here

File: sim_pipeline/RSP.py
Comment (sibirrer) [inline sim_pipeline/RSP.py:18]: please describe what this function does, also describe a bit more what are required for the butler initialization

File: sim_pipeline/RSP.py
Comment (sibirrer) [inline sim_pipeline/RSP.py:50]: dito, have a description of the definition before mentioning the parameter inputs

File: sim_pipeline/RSP.py
Comment (sibirrer) [inline sim_pipeline/RSP.py:62]: these are very hard-coded hidden (from outside of the function) parts. Perhaps it's better to have these as input options that come with default values if not specified otherwise.

File: sim_pipeline/RSP.py
Comment (sibirrer) [inline sim_pipeline/RSP.py:59]: this is not well specified. From reading the code, the output is a Table format. Please be specific

File: sim_pipeline/image_simulation.py
Comment (sibirrer) [inline sim_pipeline/image_simulation.py:68]: describe the function, also the name might not be obvious, perhaps re-defineit to galsimobj()?

File: sim_pipeline/image_simulation.py
Comment (sibirrer) [inline sim_pipeline/image_simulation.py:78]: describe function

File: sim_pipeline/image_simulation.py
Comment (sibirrer) [inline sim_pipeline/image_simulation.py:88]: describe function

File: sim_pipeline/Plots/gg_plots.py
Comment (sibirrer) [inline sim_pipeline/Plots/gg_plots.py:4]: PEP8 standards require two white spaces before a class definition. If you are using PyCharm, you can have a plugin that highlights such things

**Review by sibirrer — COMMENTED** [review]
@nkhadka21 I made some comments for you to review.
Please do not directly merge into any branch of the simulation pipeline but instead use pull request. Giving you the privilege to merge does not mean you should circumvent the pull request and reviews - thank you!

**Comment by ajshajib** [comment]
> PEP8 standards require two white spaces before a class definition. If you are using PyCharm, you can have a plugin that highlights such things

I can set up some CI functionality using pre-commit (https://pre-commit.com/) to automatically catch (and fix) PEP8 and other style-related issues. This should save time both for the coder and the reviewers. I am already using this in another project, which is working out great. If this looks useful, I can provide a PR.

**Comment by sibirrer** [comment]
> > PEP8 standards require two white spaces before a class definition. If you are using PyCharm, you can have a plugin that highlights such things
> 
> I can set up some CI functionality using pre-commit (https://pre-commit.com/) to automatically catch (and fix) PEP8 and other style-related issues. This should save time both for the coder and the reviewers. I am already using this in another project, which is working out great. If this looks useful, I can provide a PR.

Thank you @ajshajib ! Yes, that would be great. If you can make a PR directly to main with this, that will help.

**Comment by ajshajib** [comment]
A good number of PEP8/style errors thrown by `ruff` is about line length being >88 characters. 88 is (although higher than the PEP8 limit of 79) the new default adopted by current linters. But, it's possible to set a limit of our own choice. Although I prefer to stick with 88, @sibirrer or anyone else have other preferences?

File: sim_pipeline/Sources/galaxies.py
Comment (smericks) [inline sim_pipeline/Sources/galaxies.py:25]:
```suggestion
        column_names = galaxy_list.colnames
        if 'ellipticity' not in column_names:
            raise ValueError('required parameters missing in galaxy_list columns')
```

**Review by smericks — COMMENTED** [review]
- In general, make sure functions and classes have a description underneath them
- If a parameter is set to None as a default, it would be helpful to know what exactly that does somewhere in the docstring underneath (i.e. filters=None means what?)
- The Lenses folder could use some refactoring. There are some common functions / code duplication between all_lens_galaxies.py and early_type_lens_galaxies.py that could go into a base class. 
- Also there are some detached functions at the end of those files, should they be included within the class as a static method?
- In general, it could be better to call lens galaxies "deflectors" rather than "lenses", since "lenses" are the full object with all of the components.
- I'm happy to help make some of these changes either in this PR (not sure how) or in a future PR!

File: sim_pipeline/RSP.py
Comment (ajshajib) [inline sim_pipeline/RSP.py:46]:
```suggestion
    cutout_size = geom.ExtentI(num_pix, num_pix)
    #print(cutout_size)


    #Read this from the table we have at hand... 
    tract_info = skymap.find_tract(point)
    patch_info = tractInfo.findPatch(point)
    my_tract = tract_info.tract_id
    my_patch = patch_info.getSequentialIndex()
    xy = geom.PointI(tractInfo.getWcs().skyToPixel(point))
    bbox = geom.BoxI(xy - cutout_size//2, cutout_size)
    coadd_id_r = {
        'tract':my_tract, 
        'patch':my_patch,
        'band': band
    }
    coadd_cut_r = butler.get("deepCoadd", parameters={'bbox':bbox}, dataId=coadd_id_r)
```

File: sim_pipeline/RSP.py
Comment (ajshajib) [inline sim_pipeline/RSP.py:88]: Repeated code could be put in a function/method. Otherwise, the variable names should not use camelCase style.

File: sim_pipeline/RSP.py
Comment (ajshajib) [inline sim_pipeline/RSP.py:89]:
```suggestion
        lens = sharp_image(lens_class=lens_class, band=band, mag_zero_point=27, delta_pix=delta_pix, num_pix=num_pix)
```

File: sim_pipeline/RSP.py
Comment (ajshajib) [inline sim_pipeline/RSP.py:108]:
```suggestion
        point_r = wcs_r.pixelToSky(center_r)
```

File: sim_pipeline/image_simulation.py
Comment (ajshajib) [inline sim_pipeline/image_simulation.py:20]: `import` statement should be at the top of the file.

File: sim_pipeline/gg_lens_pop.py
Comment (ajshajib) [inline sim_pipeline/gg_lens_pop.py:46]:
```suggestion
        if lens_type == 'elliptical':
```

File: sim_pipeline/gg_lens_pop.py
Comment (ajshajib) [inline sim_pipeline/gg_lens_pop.py:12]:
```suggestion
    def __init__(self, lens_type='elliptical', source_type='galaxies', kwargs_deflector_cut=None,
```

File: sim_pipeline/gg_lens_pop.py
Comment (ajshajib) [inline sim_pipeline/gg_lens_pop.py:181]: Why the 1.3 value?

File: sim_pipeline/image_simulation.py
Comment (ajshajib) [inline sim_pipeline/image_simulation.py:7]: Needs one line to describe the functionality.

File: sim_pipeline/image_simulation.py
Comment (ajshajib) [inline sim_pipeline/image_simulation.py:40]: Needs one line to describe the functionality.

File: sim_pipeline/image_simulation.py
Comment (ajshajib) [inline sim_pipeline/image_simulation.py:104]: Needs one line to describe the functionality.

File: sim_pipeline/selection.py
Comment (ajshajib) [inline sim_pipeline/selection.py:3]: Needs one line to describe the functionality.

File: notebooks/lens_source_injection.ipynb
Comment (ajshajib) [inline notebooks/lens_source_injection.ipynb:1]: I don't understand the purpose of this notebook. The last cell shows that the functionality of the whole notebook can be achieved by calling `RSP.lens_injection()`, so was the purpose of the codes above it was to demonstrate how that function works? Then, I suggest putting the last cell at the top of the notebook so that a user looking for that functionality can find the info on correct usage right away. Then, the notebook should state that the codes below describe what `RSP.lens_injection()` is doing under the hood.

**Review by ajshajib — CHANGES_REQUESTED** [review]

File: notebooks/lens_source_injection.ipynb
Comment (nkhadka21) [reply notebooks/lens_source_injection.ipynb:1]: @ajshajib, you are right. Parts other than rsp.lens_injection() shows how rsp.lens_injkection() works.

**Comment by nkhadka21** [comment]
Hi @ajshajib. Now the gg-lensing branch has a complete code. So, after you review it, possibly we can merge it to the main branch!

**Comment by ajshajib** [comment]
> Hi @ajshajib. Now the gg-lensing branch has a complete code. So, after you review it, possibly we can merge it to the main branch!

@nkhadka21 Thanks a lot for the updates on the code. I see that many of the review comments from me and @smericks are unresolved. I think, all of them need to be resolved before a final round of review and then merging. Please ping me when those comments are resolved.

File: sim_pipeline/rsp.py
Comment (ajshajib) [inline sim_pipeline/rsp.py:1]: What does `rsp` stand for? I suggest using a more descriptive filename so that it's clearer what this module is about.

Comment (nkhadka21) [reply sim_pipeline/rsp.py:1]: Thanks @ajshajib !

**Comment by ajshajib** [comment]
@sibirrer, I couldn't find the coverage of the test functions. Is `coveralls` not working at the moment?

**Comment by trivialTZ** [comment]
@ajshajib @sibirrer I wonder do we need to change the `pytest --cov=sim-pipeline` to ` pytest --cov=sim_pipeline` here to make `coveralls` works? 

[pytest --cov=sim-pipeline](https://github.com/LSST-strong-lensing/sim-pipeline/blob/86a89e3248e20bb6f4d64fe1f8fb5ff4e0177d12/.github/workflows/tests.yml#L38)

File: sim_pipeline/Sources/galaxies.py
Comment (nkhadka21) [reply sim_pipeline/Sources/galaxies.py:25]: Hi @smericks , I think it has already checked whether columns_names contains ellipticity or not at the beginning.

**Comment by sibirrer** [comment]
> @sibirrer, I couldn't find the coverage of the test functions. Is `coveralls` not working at the moment?

hmm, I don't know. Sometimes it has weird flukes. We could add a makefile and @nkhadka21 can run the tests at least locally.

**Comment by sibirrer** [comment]
> @ajshajib @sibirrer I wonder do we need to change the `pytest --cov=sim-pipeline` to ` pytest --cov=sim_pipeline` here to make `coveralls` works?
> 
> [pytest --cov=sim-pipeline](https://github.com/LSST-strong-lensing/sim-pipeline/blob/86a89e3248e20bb6f4d64fe1f8fb5ff4e0177d12/.github/workflows/tests.yml#L38)

hmm, that might actually be. You can test it out (if not already, as I am behind the messages).
On another note, perhaps we should also change the name of the sim-pipeline package for better recognizability. I just did not come up with a better name and just put something in to get it going.

**Comment by ajshajib** [comment]
> > @ajshajib @sibirrer I wonder do we need to change the `pytest --cov=sim-pipeline` to ` pytest --cov=sim_pipeline` here to make `coveralls` works?
> > [pytest --cov=sim-pipeline](https://github.com/LSST-strong-lensing/sim-pipeline/blob/86a89e3248e20bb6f4d64fe1f8fb5ff4e0177d12/.github/workflows/tests.yml#L38)
> 
> hmm, that might actually be. You can test it out (if not already, as I am behind the messages). On another note, perhaps we should also change the name of the sim-pipeline package for better recognizability. I just did not come up with a better name and just put something in to get it going.

Maybe we can open a separate thread (it would be great if an owner/maintainer could open up the Discussions board for this repo) to discuss a nice name for the pipeline. As a suggestion, how about `LensIm` for the name, joining "Lens" and "Sim", pronounced as "len-sim"?

**Comment by sibirrer** [comment]
> > > @ajshajib @sibirrer I wonder do we need to change the `pytest --cov=sim-pipeline` to ` pytest --cov=sim_pipeline` here to make `coveralls` works?
> > > [pytest --cov=sim-pipeline](https://github.com/LSST-strong-lensing/sim-pipeline/blob/86a89e3248e20bb6f4d64fe1f8fb5ff4e0177d12/.github/workflows/tests.yml#L38)
> > 
> > 
> > hmm, that might actually be. You can test it out (if not already, as I am behind the messages). On another note, perhaps we should also change the name of the sim-pipeline package for better recognizability. I just did not come up with a better name and just put something in to get it going.
> 
> Maybe we can open a separate thread (it would be great if an owner/maintainer could open up the Discussions board for this repo) to discuss a nice name for the pipeline. As a suggestion, how about `lenSim` for the name, joining "Lens" and "Sim"?

@ajshajib I though you were already an owner/maintainer, since you were a project owner. I added you now also to the sim-pipeline repo

File: sim_pipeline/gg_lens_pop.py
Comment (nkhadka21) [reply sim_pipeline/gg_lens_pop.py:181]: The factor should more be like 2. So it is effectively the area within we can expect multiple images to occur, and be conservative with the area such that all possible multiply imaged configurations land in this area.

Comment (sibirrer) [reply sim_pipeline/gg_lens_pop.py:181]: I would change it to 2.5 and then comment on it as you did here @nkhadka21. Thanks a lot!

Comment (nkhadka21) [reply sim_pipeline/gg_lens_pop.py:181]: Thanks Simon. I changed it to 2.5. I will make a PR with all the changes later.

File: sim_pipeline/gg_lens_pop.py
Comment (nkhadka21) [reply sim_pipeline/gg_lens_pop.py:17]: Changed!

**Comment by nkhadka21** [comment]
Hi @ajshajib, @smericks, I worked on most of the comments. So, I think you can look at it now and feel free to make changes as you see fit!

File: notebooks/gg_lensing_tutorial.ipynb
Comment (ajshajib) [inline notebooks/gg_lensing_tutorial.ipynb:1]: Can the name of this notebook be changed to `galaxy_galaxy_lensing_tutorial.ipynb`?

File: tests/test_Lenses/test_velocity_dispersion.py
Comment (ajshajib) [inline tests/test_Lenses/test_velocity_dispersion.py:1]: The name of the folder should be changed to `test_Deflectors`.

File: tests/test_gg_lens.py
Comment (ajshajib) [inline tests/test_gg_lens.py:1]: The name of this file should be changed to `test_galaxy_galaxy_lens.py`.

**Comment by ajshajib** [comment]
Thanks a lot for your hard work @nkhadka21! Have you run `coverage` locally and checked how much of the code is covered by the test suite?

**Comment by nkhadka21** [comment]
> Thanks a lot for your hard work @nkhadka21! Have you run `coverage` locally and checked how much of the code is covered by the test suite?

I run coverage and only 54% of the code is covered in the test suite.

**Comment by ajshajib** [comment]
> I run coverage and only 54% of the code is covered in the test suite.

Ideally, we would like to have 100% coverage, unless there is a good reason for less than that.

**Comment by nkhadka21** [comment]
> > I run coverage and only 54% of the code is covered in the test suite.
> 
> Ideally, we would like to have 100% coverage, unless there is a good reason for less than that.

Yes, I agree.

**Comment by smericks** [comment]
I agree 100% coverage is ideal. Should we be flexible since this is such a large PR and we want changes merged into main in a timely manner? And then work towards 100% coverage with future commits? I'm not sure how involved it would be to do the remaining coverage?

**Comment by nkhadka21** [comment]
> I agree 100% coverage is ideal. Should we be flexible since this is such a large PR and we want changes merged into main in a timely manner? And then work towards 100% coverage with future commits? I'm not sure how involved it would be to do the remaining coverage?

Hi @smericks @ajshajib , it sounds reasonable. We can be flexible.

**Comment by ajshajib** [comment]
> I agree 100% coverage is ideal. Should we be flexible since this is such a large PR and we want changes merged into main in a timely manner? And then work towards 100% coverage with future commits? I'm not sure how involved it would be to do the remaining coverage?

I don't know if it is practically necessary to merge this PR soon, is it? I thought any user wanting to use the code could just use the `gg-lensing` branch for now. But, merging now vs. later would make a difference in the code reviewer's job. So, I can see two ways to go about it:

1. Merge this PR, open an issue requiring 100% coverage, and a future PR would close that issue. This naturally allows new reviewers.
2. Make the tests and create a PR to the `gg-lensing` branch with them. New reviewers can then be allowed to review that PR. After that PR is merged to the `gg-lensing` branch, this PR would be merged to the `main`.

@smericks is proposing Way 1, and I am fine with that as long as the plan is to increase the test coverage as the very next task. Do @sibirrer and @trivialTZ have any preferences? @nkhadka21 is indifferent, it sounds.

**Comment by nkhadka21** [comment]
> 

Hi @ajshajib and @smericks , I think currently coverage is low (only 54%). So, I think we should increase it. I will work on it and let you know.

**Comment by ajshajib** [comment]
> Hi @ajshajib and @smericks , I think currently coverage is low (only 54%). So, I think we should increase it. I will work on it and let you know.

That sounds great, @nkhadka21. It would be helpful if you could have the newly added tests as a separate PR to the `gg-lensing` branch and have it reviewed separately before merging. Then, it would be easier to go through those new additions/files, since this PR already contains quite a lot of files.

**Review by ajshajib — APPROVED** [review]
I am approving this PR contingent upon the test modules covering 100% of the code (or as close to 100% as possible), which should be added through a separate PR to the gg-lensing branch. Once the other PR is independently reviewed and merged to gg-lensing, feel free to proceed with merging this PR to main, as far as I am concerned. Thanks a lot for all the hard work from everyone involved! 🙂

**Comment by nkhadka21** [comment]
Hi @sibirrer, @ajshajib, and @smericks , now the code coverage is 73%.

**Comment by ajshajib** [comment]
I leave it up to @sibirrer to decide what test coverage would be sufficient for merging.

**Comment by sibirrer** [comment]
> I leave it up to @sibirrer to decide what test coverage would be sufficient for merging.

@ajshajib this is sufficient for me at the moment. Parts of the code need to be run on the Rubin Science Platform and are harder to test. We hope as things evolve that we can increase the test coverage and are more demanding to ourselves and others

**Comment by nkhadka21** [comment]
> > I leave it up to @sibirrer to decide what test coverage would be sufficient for merging.
> 
> @ajshajib this is sufficient for me at the moment. Parts of the code need to be run on the Rubin Science Platform and are harder to test. We hope as things evolve that we can increase the test coverage and are more demanding to ourselves and others

Ok. Let's merge it!

**Comment by smericks** [comment]
@nkhadka21 this is great! Happy to see it merged :)

**Pull request merged**
